File: src/translator/zh2en/provider/property.ts
import { DATA as POE_DATA } from "../../../data/poe/index.js";
import type { Property } from "../../../data/poe/types.js";
import { getTextSkeleton } from "../util/text.js";

const VARIABLE_PLACEHOLDER = "{0}";

export class PropertyProvider {
    private readonly zhIdx = new Map<string, Property>();
    private readonly zhSkeletonIdx = new Map<string, Property[]>();

    constructor() {
        for (const p of POE_DATA.properties) {
            const zh = p.zh;
            this.zhIdx.set(zh, p);

            if (zh.includes(VARIABLE_PLACEHOLDER)) {
                const zhSkeleton = getTextSkeleton(zh);
                const value = this.zhSkeletonIdx.get(zhSkeleton);
                if (value) {
                    value.push(p);
                } else {
                    this.zhSkeletonIdx.set(getTextSkeleton(zh), [p]);
                }
            }
        }
    }

    provideByZh(zh: string): Property | undefined {
        return this.zhIdx.get(zh);
    }

    provideVariablePropertiesByZhSkeleton(
        skeleton: string,
    ): Property[] | undefined {
        return this.zhSkeletonIdx.get(skeleton);
    }
}

File: go/translator/zh2en/provider/property.go
package provider

import (
	"strings"

	"github.com/cn-poe-community/cn-poe-utils/go/data/poe"
	"github.com/cn-poe-community/cn-poe-utils/go/translator/zh2en/util"
)

const VARIABLE_PLACEHOLDER = "{0}"

type PropertyProvider struct {
	zhIdx         map[string]*poe.Property
	zhSkeletonIdx map[string][]*poe.Property
}

func NewPropertyProvider() *PropertyProvider {
	zhIdx := make(map[string]*poe.Property)
	zhSkeletonIdx := make(map[string][]*poe.Property)

	for i := range poe.DATA.Properties {
		prop := &poe.DATA.Properties[i]

		if strings.Contains(prop.Zh, VARIABLE_PLACEHOLDER) {
			zhSkeletonIdx[util.GetTextSkeleton(prop.Zh)] =
				append(zhSkeletonIdx[util.GetTextSkeleton(prop.Zh)], prop)
		} else {
			zhIdx[prop.Zh] = prop
		}
	}
	return &PropertyProvider{
		zhIdx:         zhIdx,
		zhSkeletonIdx: zhSkeletonIdx,
	}
}

func (p *PropertyProvider) ProvideByZh(zh string) *poe.Property {
	if prop, ok := p.zhIdx[zh]; ok {
		return prop
	}
	return nil
}

func (p *PropertyProvider) ProvideVariablePropertiesByZhSkeleton(skeleton string) []*poe.Property {
	if props, ok := p.zhSkeletonIdx[skeleton]; ok {
		return props
	}
	return nil
}
